spec/ffi/embed-test/ext/embed.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2017 Thomas Martitz <kugel@rockbox.org>
+ * Copyright (C) 2008-2017, Ruby FFI project contributors
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Ruby FFI project nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <dlfcn.h>
+
+#include <ruby.h>
+
+const char script[] =
+"require 'rubygems'\n"
+"require 'ffi'\n"
+"module LibWrap\n"
+"  extend FFI::Library\n"
+"  ffi_lib FFI::CURRENT_PROCESS\n"
+"  callback :completion_function, [:string, :long, :uint8], :void\n"
+"  attach_function :do_work, [:pointer, :completion_function], :int\n"
+"  Callback = Proc.new do |buf_ptr, count, code|\n"
+"    nil\n"
+"  end\n"
+"end\n"
+"\n"
+"LibWrap.do_work(\"test\", LibWrap::Callback)\n";
+
+typedef void completion_function(const char *buffer, long count, unsigned char code);
+
+extern int do_work(const char *buffer, completion_function *);
+
+static completion_function *ruby_func;
+
+int do_work(const char *buffer, completion_function *fn)
+{
+  /* Calling fn directly here works */
+  ruby_func = fn;
+  return 0;
+}
+
+static VALUE testfunc(VALUE args);
+
+void Init_embed_test(void)
+{
+  VALUE mod = rb_define_module("EmbedTest");
+  rb_define_module_function(mod, "testfunc", testfunc, 0);
+}
+
+static VALUE testfunc(VALUE self)
+{
+  int state = 0;
+  VALUE ret;
+
+  rb_eval_string_protect(script, &state);
```

Comment:
> Is RBX failing because of incorrect behavior in FFI when running under RBX?

It's failing because RBX doesn't know `rb_eval_string_protect()` but only `rb_eval_string()`. However the `state` is important to verify that the embedded script worked. So the program has to be changed, so that the success state is transferred in some other way (`rb_rescue()` or similar).